website/server/controllers/api-v4/members.js

```diff
@@ -23,25 +25,64 @@ api.getUsernameAutocompletes = {
       return;
     }
 
-    let query = {'auth.local.lowerCaseUsername': {$regex: `^${username}.*`}, 'flags.verifiedUsername': true, 'preferences.searchableUsername': {$ne: false}};
+    let commonQuery = {
+      'flags.verifiedUsername': true,
+      'auth.blocked': {$ne: true},
+      'flags.chatRevoked': {$ne: true},
+      'auth.local.lowerCaseUsername': {$regex: `^${username}.*`},
+    };
+    let query = Object.assign({}, commonQuery);
 
     let context = req.query.context;
-    let id = req.query.id;
-    if (context && id) {
-      if (context === 'party') {
-        query['party._id'] = res.locals.user.party._id;
-      } else if (context === 'privateGuild') {
-        if (res.locals.user.guilds.includes(id)) {
-          query.guilds = id;
-        }
+    let groupID = req.query.id;
+
+    let members = [];
+    let isPublicSpace = true;
+    if (context && groupID) {
+      if (context === 'party' && res.locals.user.party._id === groupID) {
+        query['party._id'] = groupID;
+        isPublicSpace = false;
+      } else if (context === 'privateGuild' && res.locals.user.guilds.includes(groupID)) {
+        query.guilds = groupID;
+        isPublicSpace = false;
+      } else if (context !== 'publicGuild' && context !== 'tavern') {
+        res.respond(200, []);
+        return;
       }
+
+      let recentChats = await Chat
+        .find({groupId: groupID, username: {$regex: `^${username}.*`}})
+        .select(['uuid'])
+        .sort({timestamp: -1})
+        .limit(200)
+        .exec();
+      let recentChatters = _.uniq(recentChats.map((message) => {
+        return message.uuid;
+      }));
+
+      let recentChatQuery = Object.assign({}, commonQuery);
+      recentChatQuery._id = {$in: recentChatters};
+      query._id = {$nin: recentChatters};
+      members = await User
+        .find(recentChatQuery)
+        .select(['profile.name', 'contributor', 'auth.local.username'])
+        .sort({'auth.timestamps.loggedin': -1})
+        .limit(5 - members.length)
```

Comment:
member before this query isn't always an empty array? So members.length will always be 0, or not?